### PR TITLE
Start some suggestions around ShellContext

### LINKED ISSUE #1835: Lucene writer locking issue
Repro (maybe not all time but most of the time on my machine)

- Add to the blog recipe `OrchardCore.Lucene.Worker`.
 
- Run a blog setup.

- Lucene writer fails on time out to obtain a lock.

Then using #1669 solved the issue.

When concurrent calls are done on `GetOrCreateShellContext()`, multiple shell contexts are created and not added to the dictionary, this is the GC which disposes them through their finalizers. See comments in #1646 where i did apache tests and the issues fixed by using a **dictionnary of lazy values**.

A problem is when a service is an `IDisposable` it is disposed later by the GC. With `LuceneIndexManager` the result is that the writers are not disposed and then the lock is not freed immediately.

Here we have this case during the setup, at some point while the shell is recreated through the factory lambda, the lucene worker through its background task also call `GetOrCreateShellContext()`.

Note: Using #1669, i could see that `LuceneIndexManager` was correctly disposed.

### PR DESCRIPTION
**Pull request by jtkech**
#Related to my last comment in #1646, to be continued.

Fixes #1835

### CODE REVIEW

**Comment by jtkech** [comment]
So, because now the shells collection can contain a dependent which could be disposed and then with a null settings, in my last commit `OC.Tenants` uses `_shellSettingsManager.LoadSettings()`, this in place of first using `_orchardHost.ListShellContexts()` to retrieve their settings.

Maybe not good because here we always triggers the `IShellSettingsConfigurationProvider`, but `ListShellContexts()` might not reflect all tenants. On an another side we only do it through the admin.

Finally, we could do the same in the `hostedService` branch where i also use `ListShellContexts()`, otherwise the bgTasks of a disposed dependent shell could not be run. But here maybe bad to always use `.LoadSettings()` in the background. Let me know.

Hmm, just thought about another solution, when we dispose a `ShellContext`, maybe we could not set its `Settings` property to null, so that as long as a shell is still in the collection (or referenced somewhere else) we could retrieve its setting and then be able to call `GetOrCreateShellContext()` on it.

This could also help the work in the background where i already did a wrapper class to wrap a shell with its `Settings.Name`, this to retrieve the name even if the shell is disposed, this to release its schedulers.

So, let me know if there is another reason to set `Settings` to null when disposing a `ShellContext`, e.g if it is just to tell to the GC to collect it ealier, i don't think it could help.

I just commited the changes related to `OC.Tenants` to show what we would need to do. **Then i will revert it and do this suggestion** related to not set `Settings` to null when disposing a `ShellContext`.

File: src/OrchardCore.Modules/OrchardCore.OpenId/Configuration/OpenIdValidationConfiguration.cs [hunk 1]
```diff
@@ -78,6 +79,11 @@ public void Configure(AuthenticationOptions options)
             // this using block won't be disposed until the service scope itself is released.
             using (var scope = CreateTenantScope(settings.Tenant))
             {
+                if (scope == null)
```
Comment (PinpointTownes) [inline src/OrchardCore.Modules/OrchardCore.OpenId/Configuration/OpenIdValidationConfiguration.cs]: Doing that is problematic as it will leave the tenant that uses the validation feature in a state where it can't properly validate incoming tokens.

If this is the pattern we adopt, we'll probably have to replace the `return` by a signal telling the module middleware to "restart" the tenant (by creating a new `ShellContext` with a new service provider). E.g something like `throw new ShellFatalException()`.

File: src/OrchardCore.Modules/OrchardCore.OpenId/Configuration/OpenIdValidationConfiguration.cs [hunk 2]
```diff
@@ -208,19 +219,23 @@ public void Configure(string name, OAuthValidationOptions options)
             if (!string.IsNullOrEmpty(settings.Tenant) &&
                 !string.Equals(settings.Tenant, _shellSettings.Name, StringComparison.Ordinal))
             {
-                var context = _shellHost.GetOrCreateShellContext(_shellSettingsManager.GetSettings(settings.Tenant));
-                using (var scope = context.EnterServiceScope())
+                var shellSettings = _shellSettingsManager.GetSettings(settings.Tenant);
+
+                if (shellSettings.State == TenantState.Running)
```
Comment (PinpointTownes) [inline src/OrchardCore.Modules/OrchardCore.OpenId/Configuration/OpenIdValidationConfiguration.cs]: This is also problematic. Ideally, `_shellHost.EnterServiceScope` should take care of starting the server tenant if it's stopped at the time we reach this code path.

File: src/OrchardCore.Modules/OrchardCore.OpenId/Drivers/OpenIdValidationSettingsDisplayDriver.cs [hunk 3]
```diff
@@ -71,7 +73,7 @@ public class OpenIdValidationSettingsDisplayDriver : SectionDisplayDriver<ISite,
                 model.Audience = settings.Audience;
                 model.Tenant = settings.Tenant;
 
-                model.AvailableTenants = (from tenant in _shellSettingsManager.LoadSettings().AsParallel()
+                model.AvailableTenants = (from tenant in _shellSettingsManager.LoadSettings().Where(s => s.State == TenantState.Running).AsParallel()
                                           let provider = _shellHost.GetOrCreateShellContext(tenant).ServiceProvider
```
Comment (PinpointTownes) [inline src/OrchardCore.Modules/OrchardCore.OpenId/Drivers/OpenIdValidationSettingsDisplayDriver.cs]: We should probably remove this LINQ query and replace it by a `foreach` to be able to use the `EnterServiceScope` API in a `using` block.

File: src/OrchardCore/OrchardCore.Environment.Shell/ShellHost.cs [hunk 4]
```diff
@@ -89,6 +89,38 @@ public ShellContext GetOrCreateShellContext(ShellSettings settings)
...
         }
 
+        public IServiceScope EnterServiceScope(ShellSettings settings)
+        {
+            return EnterServiceScope(settings, out var context);
+        }
+
+        public IServiceScope EnterServiceScope(ShellSettings settings, out ShellContext context)
+        {
+            context = GetOrCreateShellContext(settings);
```
Comment (PinpointTownes) [inline src/OrchardCore/OrchardCore.Environment.Shell/ShellHost.cs]: I suppose the `ShellContext` might be released immediately after this method is called (and before `TryEnterServiceScope` has a chance to be invoked).

Should `ShellHost` use good old `locks` for all its operations?

Comment (jtkech) [reply src/OrchardCore/OrchardCore.Environment.Shell/ShellHost.cs]: Still working on it, just restarted but always need time to re-put all of this into my mind ;)

But the idea is to keep the fastest path in most of the cases. Then, yes the shell context might be released and `TryEnterServiceScope()` would return false (but without failing).

Then, if so we remove the shell from the concurrent dictionary and re-create it, as we now do in `GetOrCreateShellContext()` (see note 1). This is here that the atomic operation occurs in the valueFactory of the lazy value and where we create the scope.

But we will never be able to do this if the tenant is already **disabled**, the created shell would be not released but with a null service provider (see note 2), that's why we check this just before.

**Note 1**: Before, when a shell was released it was removed from the concurrent dictionary, but now a dependent shell might be released and still in the dictionary. Now in `GetOrCreateShellContext()`, we take care of this by removing and re-creating it through the value factory of the lazy value.

**Note 2**: When you get a shell settings, as long as it is not hold by an active shell context, it's state is not changed, but if it is already **disabled** `GetOrCreateShellContext()` will return a context wich is not released but with a null `ServiceProvider`.

**Note 3**: But if you already are in an active scope and the shell is **disabled** after, it is marked as released and removed from the shells dico. But you can still ends your scope with a non null provider. But on the next `GetOrCreateShellContext()` the shell will be re-created as described above.

File: src/OrchardCore.Modules/OrchardCore.OpenId/Drivers/OpenIdValidationSettingsDisplayDriver.cs (hunk 3 above)
Comment (jtkech) [reply src/OrchardCore.Modules/OrchardCore.OpenId/Drivers/OpenIdValidationSettingsDisplayDriver.cs]: Okay cool, i will do it after reviewing the other points.

File: src/OrchardCore.Modules/OrchardCore.OpenId/Configuration/OpenIdValidationConfiguration.cs (hunk 2 above)
Comment (jtkech) [reply src/OrchardCore.Modules/OrchardCore.OpenId/Configuration/OpenIdValidationConfiguration.cs]: This is what it does (see my other comment below), but:

I think that here, when you reference another tenant e.g it can't be `Uninitialized` or `Initializing` meaning it has not been created or setup. So, the only problem would be if it has been **disabled**.

When you get a shell settings, as long as it is not hold by an active shell context, it's state is not changed, but if it is already **disabled** `GetOrCreateShellContext()` will return a context wich is not released but with a null `ServiceProvider`. And you will never be able to create a service scope.

Note: If you already are in an active scope and the shell is disabled after, it is marked as released and removed from the shells dico. But you can still ends your scope with a non null provider. But on the next `GetOrCreateShellContext()` the shell will be re-created as described above.

Note: The `OC.Tenants` index page lists all shell contexts including disabled tenants.

In summary, maybe here the only problem is if the tenant is already disabled. And in this case you will never be able to create a service scope. Let me know what is better to do in this case.

File: src/OrchardCore.Modules/OrchardCore.OpenId/Configuration/OpenIdValidationConfiguration.cs (hunk 1 above)
Comment (jtkech) [reply src/OrchardCore.Modules/OrchardCore.OpenId/Configuration/OpenIdValidationConfiguration.cs]: As said below, i think the only use case is when the tenant is already disabled and this is what i'm ckecking in `CreateTenantScope()`. But i will remove the check in `CreateTenantScope()` because currently i do it directly in `EnterServiceScope()` which returns null if the shell is disabled.

When we are in an active scope of a given tenant that's ok. But here the singularity is that we create a scope from another tenant, so at some point we need to check if it is not disabled.

So, maybe not a pattern to adopt, just a temporary one. But it depends on what we want to do in such a case. Hmm, maybe we can't throw a non catched exception if the result is that we can't restart the application.

As a side note, doing some tests on `OpenId`, i'm a newbie around this ;), i was in some cases where i was not able to restart the application, only by doing a new setup. This when it was telling to me that there is a missing certificate, and also when there was a duplicate certificate.

**Comment by jtkech** [comment]
@PinpointTownes i began to react to your review.

So, see my comments (even outdated ones), particularly when a tenant is disabled so that you can create a shell which is not released but with a null service provider on wich you can't create a service scope.

That said, because in `OpenIdValidationService.ValidateSettingsAsync()` i also check the tenant state (now if it is disabled to be clearer), i think it is no more needed to check it in the other places we talked about, because `ValidateSettingsAsync()` is always called before. Hmm, at some point i remember i was not able to put some breakpoints, maybe after this new checking.

So now there is only one place where we check `TenantState.Disabled`, in `ValidateSettingsAsync()` line 132, and where we do the following. Let me know if it is ok.

    results.Add(new ValidationResult(T["The specified tenant is disabled."]));

Also, remember the use case i had where i was not able to restart the app without doing a new setup. This when it was telling about a missing or a duplicate certificate.

**Comment by jtkech** [comment]
@PinpointTownes @sebastienros for infos here my progress for testing.

- So, here i suggest that when we disable a shell, if it is still in an active scope, it is not removed from the shells dictionary and it will be released and then disposed by its last scope. But it is still removed from the running shell table, so that it is no more served by subsequent requests.

    Doing so, e.g when we disable a shell in the middle of a request scope, before ending this scope now we can also enter in a new inner scope without failing. Otherwise, the disabed shell would be recreated and would have a null service provider.

    That said, when we enter in a new scope **from another tenant** (or in the background) we still need to check if the returned scope is null, meaning that the tenant is disabled. So i re-added some checks in the `OpenId` module to be reviewed. But seems to be ok everywhere else.

- About entering in a scope atomically, i think i was not covering all cases. Using lazy initializers is ok to prevent the factory of a shell to be executed multiple times. And i was using the same pattern to enter in a scope by using another factory which was creating a shell and then also creating a new scope.

    But doing so there are 2 paths to create a new shell context with 2 different factories. And, **if we want to keep a fast path**, it seems that we would have the same kind of problem even with a regular lock.

    So, here we use the same `GetOrCreateShellContext()` (which recreates a shell if released). Then we try to enter in a scope (without failing), then if the shell has been released just before (e.g a feature has changed), we just retry it recursively. Let me know if it is ok and / or better to limit the recursive calls.

**Comment by jtkech** [comment]
@sebastienros just for infos.

Doing some setup tests on the tenants branch i had a lucene writer locking issue. Then using this PR solved the issue, see #1835 for more details.

**Comment by jtkech** [comment]
@PinpointTownes 

So now, when depending on another tenant, we throw an `ApplicationException` if the "other" tenant is disabled. In this case, we need to re-enable the "other" tenant before rebuilding the pipeline of the "current" tenant by doing another request on it.

File: src/OrchardCore/OrchardCore.Environment.Shell/ShellHost.cs [hunk 5]
```diff
@@ -70,23 +71,92 @@ public void Initialize()
...
+        {
+            IServiceScope scope = null;
+
+            // The normal path is to return an already created context and pass a null scope.
+            // Then the caller can use 'TryEnterServiceScope' to get a non null tracked scope.
+
+            // If the context is released, the usage is to remove it from the dictionary and then recall
+            // this method, so that a tracked scope is created atomically through the lazy value factory.
+
+            var shell = _shellContexts.GetOrAdd(settings.Name, new Lazy<ShellContext>(() =>
```
Comment (sebastienros) [inline src/OrchardCore/OrchardCore.Environment.Shell/ShellHost.cs]: I'd prefer to use a double lock here.

Comment (jtkech) [reply src/OrchardCore/OrchardCore.Environment.Shell/ShellHost.cs]: Okay, but i need to put my mind back on this PR.

Hmm, i think you mean in place of using lazy values, so that we don't use any lazy value ...
